src/dsl/verifier/proxy/types.ts

```diff
@@ -1,20 +1,50 @@
 import express from 'express';
 import { LogLevel } from '../../options';
+import { JsonMap, AnyJson } from '../../../common/jsonTypes';
 
-export interface StateHandler {
-  [name: string]: () => Promise<unknown>;
+export type Hook = () => Promise<unknown>;
+
+/*
+ * State handlers map a state description to a function
+ * that can setup the provider state
+ */
+export interface StateHandlers {
+  [name: string]: StateHandlerV3;
 }
 
+/*
+ * Incoming V2 provider state shape
+ */
 export interface ProviderState {
   states?: [string];
 }
 
-export type Hook = () => Promise<unknown>;
+/*
+ * Incoming V3 provider state shape
+ */
+export interface ProviderStateV3 {
+  action: StateAction;
+  params: JsonMap;
+  state: string;
+}
+
+/**
+ * Specifies whether the state handler being setup or shutdown
+ */
+export type StateAction = 'setup' | 'teardown';
```

Comment:
I like that, I think it's worth doing now.

Reply:
So I've implemented that suggestion. 

One functional detail I'd like to clarify, is that if you only provide a `StateFunc` (that is, you don't care about setup/teardown) should we invoke it on the setup AND teardown phase? I'm assuming not, and have currently implemented it so that it will _only_ fire on setup. i.e. basically the same behaviour as it does today, except it also optionally takes parameters.

If you want the lifecycle hooks, you should use the variation that has two functions.

Does this seem reasonable to you? It certainly made the examples simpler once I did that, because they no longer needed a guard - so what I'm saying is, nice!